File: packages/format/src/av1/bit-reader.ts
import { FormatError } from "../errors.js";

/** Bounded MSB-first AV1 syntax reader. */
export class Av1BitReader {
  readonly #bytes: Uint8Array;
  readonly #path: string;
  #bitOffset = 0;

  public constructor(bytes: Uint8Array, path: string) {
    this.#bytes = bytes;
    this.#path = path;
  }

  public get bitOffset(): number {
    return this.#bitOffset;
  }

  public get bitsRemaining(): number {
    return this.#bytes.byteLength * 8 - this.#bitOffset;
  }

  public readBit(label: string): boolean {
    if (this.#bitOffset >= this.#bytes.byteLength * 8) {
      this.#fail(`truncated ${label}`);
    }
    const byte = this.#bytes[Math.floor(this.#bitOffset / 8)];
    if (byte === undefined) this.#fail(`truncated ${label}`);
    const shift = 7 - (this.#bitOffset % 8);
    this.#bitOffset += 1;
    return ((byte >> shift) & 1) === 1;
  }

  public readBits(width: number, label: string): number {
    if (!Number.isInteger(width) || width < 0 || width > 32) {
      this.#fail(`invalid bit width for ${label}`);
    }
    if (this.bitsRemaining < width) this.#fail(`truncated ${label}`);
    let value = 0;
    for (let index = 0; index < width; index += 1) {
      value = value * 2 + (this.readBit(label) ? 1 : 0);
    }
    return value;
  }

  public readTrailingBits(): void {
    if (!this.readBit("trailing_one_bit")) {
      this.#fail("trailing_one_bit must equal one");
    }
    while (this.bitsRemaining > 0) {
      if (this.readBit("trailing_zero_bit")) {
        this.#fail("trailing_zero_bit must equal zero");
      }
    }
  }

  #fail(message: string): never {
    throw new FormatError("PROFILE_INVALID", `AV1 ${message}`, {
      path: this.#path,
      offset: Math.floor(this.#bitOffset / 8)
    });
  }
}

File: flutter/packages/aval_format/lib/src/av1/bit_reader.dart
/// Bounded MSB-first AV1 syntax reader.
///
/// Dart port of `packages/format/src/av1/bit-reader.ts`.
library;

import 'dart:typed_data';

import '../errors.dart';

/// Bounded MSB-first AV1 syntax reader.
class Av1BitReader {
  Av1BitReader(this._bytes, this._path);

  final Uint8List _bytes;
  final String _path;
  int _bitOffset = 0;

  int get bitOffset => _bitOffset;

  int get bitsRemaining => _bytes.length * 8 - _bitOffset;

  bool readBit(String label) {
    if (_bitOffset >= _bytes.length * 8) {
      _fail('truncated $label');
    }
    final byte = _bytes[_bitOffset ~/ 8];
    final shift = 7 - (_bitOffset % 8);
    _bitOffset += 1;
    return ((byte >> shift) & 1) == 1;
  }

  int readBits(int width, String label) {
    if (width < 0 || width > 32) {
      _fail('invalid bit width for $label');
    }
    if (bitsRemaining < width) _fail('truncated $label');
    var value = 0;
    for (var index = 0; index < width; index += 1) {
      value = value * 2 + (readBit(label) ? 1 : 0);
    }
    return value;
  }

  void readTrailingBits() {
    if (!readBit('trailing_one_bit')) {
      _fail('trailing_one_bit must equal one');
    }
    while (bitsRemaining > 0) {
      if (readBit('trailing_zero_bit')) {
        _fail('trailing_zero_bit must equal zero');
      }
    }
  }

  Never _fail(String message) {
    throw FormatError(
      FormatErrorCode.profileInvalid,
      'AV1 $message',
      FormatErrorDetails(path: _path, offset: _bitOffset ~/ 8),
    );
  }
}
